### PR TITLE
Mat

### PR DESCRIPTION
**Pull request by connorjward**

### CODE REVIEW

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 1]
```diff
@@ -1481,7 +2167,42 @@ cdef class Mat(Object):
 
     # matrix-matrix product
 
-    def matMult(self, Mat mat, Mat result=None, fill=None):
+    def matMult(
+        self,
+        Mat mat,
+        result: Mat | None = None,
```
Comment (connorjward) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2173]:
```suggestion
        Mat result=None,
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 2]
```diff
@@ -204,6 +204,15 @@ class MatSORType(object):
 # --------------------------------------------------------------------
 
 cdef class Mat(Object):
+    """Matrix object.
+
+    Mat is described in the `PETSc manual <petsc:manual/ts>`.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:209]:
```suggestion
    Mat is described in the `PETSc manual <petsc:manual/mat>`.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 3]
```diff
@@ -296,28 +305,146 @@ cdef class Mat(Object):
...
+        petsc_options, Mat.load, petsc.MatView
+
+        """
         cdef PetscViewer vwr = NULL
         if viewer is not None: vwr = viewer.vwr
         CHKERR( MatView(self.mat, vwr) )
 
-    def destroy(self):
+    def destroy(self) -> Self:
+        """Destroy the matrix and free its memory.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:337]:
```suggestion
        """Destroy the matrix.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 4]
```diff
@@ -296,28 +305,146 @@ cdef class Mat(Object):
...
+
+        Collective.
+
+        See Also
+        --------
+        Mat.create, petsc.MatDestroy
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:343]: class references do not need to be of type `Mat.xxx`. Just use `xxx`

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 5]
```diff
@@ -296,28 +305,146 @@ cdef class Mat(Object):
...
+
+        """
         CHKERR( MatDestroy(&self.mat) )
         return self
 
-    def create(self, comm=None):
+    def create(self, comm: Comm | None = None) -> Self:
+        """Create the matrix.
+
+        Once created, the user should call `Mat.setType` or
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:352]: @dalcinl We should probably add a short man page on the pattern

```
create
setType
setSizes
setFromOptions
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 6]
```diff
@@ -327,28 +454,123 @@ cdef class Mat(Object):
...
+        Mat.getVecType, petsc.MatSetVecType
+
+        """
         cdef PetscVecType cval = NULL
         vec_type = str2bytes(vec_type, &cval)
         CHKERR( MatSetVecType(self.mat, cval) )
 
-    def getVecType(self):
+    def getVecType(self) -> str:
+        """Return the vector type the matrix returns with `Mat.createVecs`.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:521]:
```suggestion
        """Return the vector type used by the matrix.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 7]
```diff
@@ -327,28 +454,123 @@ cdef class Mat(Object):
...
+        csr: CSRIndicesType | None = None,
+        comm: Comm | None = None,
+    ) -> Self:
+        """Create a sparse unblocked matrix, optionally preallocating.
+
+        To preallocate the matrix the user can either pass ``nnz`` or ``csr``
+        describing the sparsity. If neither is set then preallocation will not
+        occur. Consult the `PETSc manual <petsc:sec_matsparse>` for
+        more information.
+
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:550]: extra blank line

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 8]
```diff
@@ -455,7 +836,32 @@ cdef class Mat(Object):
...
+        """Create a dense matrix.
+
+        Collective.
+
+        Parameters
+        ----------
+        size, bsize
+            Matrix size parameters. See `Mat.createAIJ` for usage information.
+        array
+            Matrix data. If `None` then the matrix will not be preallocated.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:855]: This is wrong I guess. If None, the data is internally allocated by petsc, while if it is nonzero it uses userdefined memory. Please double check

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 9]
```diff
@@ -505,7 +927,28 @@ cdef class Mat(Object):
         PetscCLEAR(self.obj); self.mat = newmat
         return self
 
-    def setPreallocationDense(self, array):
+    def setPreallocationDense(self, array: Sequence[Scalar]) -> Self:
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:930]: Maybe `ArrayScalar`? @dalcinl ?

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 10]
```diff
@@ -515,39 +958,179 @@ cdef class Mat(Object):
...
+
+        """
         cdef PetscMat newmat = NULL
         CHKERR( MatCreateNormal(mat.mat, &newmat) )
         PetscCLEAR(self.obj); self.mat = newmat
         return self
 
-    def createTranspose(self, Mat mat):
+    def createTranspose(self, Mat mat) -> Self:
+        """Create a virtual matrix transpose that behaves like Aᵀ.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:1013]:
```suggestion
        """Create a matrix object that behaves like Aᵀ.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 11]
```diff
@@ -515,39 +958,179 @@ cdef class Mat(Object):
...
+
+        """
         cdef PetscMat newmat = NULL
         CHKERR( MatCreateTranspose(mat.mat, &newmat) )
         PetscCLEAR(self.obj); self.mat = newmat
         return self
 
-    def createNormalHermitian(self, Mat mat):
+    def createNormalHermitian(self, Mat mat) -> Self:
+        """Create a matrix representing (A*)ᵀA.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:1040]: No hermitian superscript in unicode?

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 12]
```diff
@@ -515,39 +958,179 @@ cdef class Mat(Object):
...
+
+        """
         cdef PetscMat newmat = NULL
         CHKERR( MatCreateNormalHermitian(mat.mat, &newmat) )
         PetscCLEAR(self.obj); self.mat = newmat
         return self
 
-    def createHermitianTranspose(self, Mat mat):
+    def createHermitianTranspose(self, Mat mat) -> Self:
+        """Create a virtual matrix transpose that behaves like (A*)ᵀ.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:1068]:
```suggestion
        """Create a matrix that behaves like (A*)ᵀ.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 13]
```diff
@@ -560,14 +1143,68 @@ cdef class Mat(Object):
         PetscCLEAR(self.obj); self.mat = newmat
         return self
 
-    def createSubMatrixVirtual(self, Mat A, IS isrow, IS iscol=None):
+    def createSubMatrixVirtual(self, Mat A, IS isrow, IS iscol=None) -> Self:
+        """Create a virtual matrix that acts as a submatrix.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:1147]:
```suggestion
        """Create a matrix that behaves as a submatrix.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 14]
```diff
@@ -1481,7 +2162,42 @@ cdef class Mat(Object):
...
-    def matMult(self, Mat mat, Mat result=None, fill=None):
+    def matMult(
+        self,
+        Mat mat,
+        Mat result=None,
+        fill: float | None = None
+    ) -> Mat:
+        """Performs matrix-matrix multiplication C=AB.
+
+        Neighbor-wise collective.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2173]: I think we agreed on Collective, Not collective and (possibly) Logically collective.

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 15]
```diff
@@ -1503,7 +2254,42 @@ cdef class Mat(Object):
...
+            Expected fill as ratio of nnz(C)/(nnz(A) + nnz(B)), use
+            ``None`` if you do not have a good estimate. If the
+            result is a dense matrix this is irrelevant.
+
+        Returns
+        -------
+        result: Mat
+            The resultant product matrix C.
+
+        Notes
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2283]: add a note on the fact that  MAT_REUSE_MATRIX is used if the result is provided as an argument. Here and in all other MatMat operations

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 16]
```diff
@@ -1525,7 +2350,53 @@ cdef class Mat(Object):
...
+        -----
+        To determine the correct fill value, run with -info and search
+        for the string "Fill ratio" to see the value actually needed.
+
+        An alternative approach to this function is to use
+        `petsc.MatProductCreate` and set the desired options before the
+        computation is done.
+
+        This routine is currently only implemented for pairs of
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2388]: Too much information

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 17]
```diff
@@ -1820,12 +2962,30 @@ cdef class Mat(Object):
...
+        petsc.MatNestGetSize
+
+        """
         cdef PetscInt nrows, ncols
         CHKERR( MatNestGetSize(self.mat, &nrows, &ncols) )
         return toInt(nrows), toInt(ncols)
 
-    def getNestISs(self):
+    def getNestISs(self) -> tuple[list[IS], list[IS]]:
+        """Return the `IS` sets partitioning the row and column spaces.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2980]:
```suggestion
        """Return the index sets representing the row and column spaces.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 18]
```diff
@@ -1837,7 +2997,16 @@ cdef class Mat(Object):
         cdef object isetscols = [ref_IS(ciscols[i]) for i from 0 <= i < ncols]
         return isetsrows, isetscols
 
-    def getNestLocalISs(self):
+    def getNestLocalISs(self) -> tuple[list[IS], list[IS]]:
+        """Return the local `IS` sets partitioning the row and column spaces.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:3001]:
```suggestion
        """Return the local index sets representing the row and column spaces.
```

**Review by stefanozampini — CHANGES_REQUESTED** [review]

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx (hunk 14 above)
Comment (nocollier) [reply src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2173]: Neighborwise collective was added after you left, but we agreed on the American version as the other docs are written this way. I will fix.

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 19]
```diff
@@ -1481,7 +2162,42 @@ cdef class Mat(Object):
...
+        """Performs matrix-matrix multiplication C=AB.
+
+        Neighborwise collective.
+
+        Parameters
+        ----------
+        mat
+            The right hand matrix B.
+        result
+            The optional resultant matrix C, uses ``MAT_REUSE_MATRIX``.
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2180]:
```suggestion
            The optional resultant matrix C. When `None`, a new matrix is created, and ``MAT_INITIAL_MATRIX`` is used.
            When C is not `None`, the matrix is reused with ``MAT_REUSE_MATRIX``
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx [hunk 20]
```diff
@@ -1481,7 +2162,42 @@ cdef class Mat(Object):
...
+            The resultant product matrix C.
+
+        Notes
+        -----
+        To determine the correct fill value, run with -info and search
+        for the string "Fill ratio" to see the value actually needed.
+
+        See also
+        --------
+        petsc.MatMatMult
```
Comment (stefanozampini) [inline src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2198]:
```suggestion
        petsc.MatMatMult, petsc.MatReuse
```

Comment (stefanozampini) [reply src/binding/petsc4py/src/petsc4py/PETSc/Mat.pyx:2198]: And in the others too. Thanks @nocollier

**Review by stefanozampini — APPROVED** [review]

**Pull request merged**